### PR TITLE
Add TextWriter.CreateBroadcasting

### LINKED ISSUE #93623: [API Proposal]: TextWriter.Aggregate
UPDATED PROPOSAL 1/3/2024:
```diff
public abstract partial class TextWriter
{
+    public static TextWriter Aggregate(ReadOnlySpan<TextWriter> writers);
}
```

Open questions:
- Is this the right name?
- Should there be an optional parameter for leaveOpen?
- Do we _need_ an array-based overload? params?

### Background and motivation

It's common in simple command-line apps/tools to want to write output to multiple places, e.g. to the console and to a file.

### API Proposal

```csharp
namespace System.IO

public sealed class TeeTextWriter : TextWriter
{
    public TeeTextWriter(params TextWriter[] writers);
    ... // overrides of all TextWriter methods
}
```

### API Usage

```csharp
using StreamWriter fileWriter = new StreamWriter("...");
var writer = new TeeTextWriter(fileWriter, Console.Out);
foreach (string s in strings)
    writer.WriteLine($"The value is {s}");
```

### Alternative Designs

_No response_

### Risks

_No response_

### PR DESCRIPTION
**Pull request by stephentoub**
Closes https://github.com/dotnet/runtime/issues/93623

### CODE REVIEW

**Comment by skyoxZ** [comment]
Does the behavior or the method name still have a chance to change? 

```c#
/// The resulting instance will delegate each operation to each of the writers in <paramref name="writers"/>.
/// For example, calling <see cref="Write(char)"/> will write the specified char to each writer, one after the
/// other. The writers will be written to in the same order as they are specified in <paramref name="writers"/>.
/// An exception from the operation on one writer will prevent the operation from being performed on subsequent writers.
```

For this behavior, `Chain` seems to be a better name for me. And for the name of `broadcasting` I would expect the behavior of `broadcasting.WriteAsync` to be something like `Task.WhenAll`.

File: src/libraries/System.Private.CoreLib/src/System/IO/TextWriter.cs [hunk 1]
```diff
@@ -1018,5 +1021,508 @@ public override Task FlushAsync(CancellationToken cancellationToken)
...
+                Debug.Assert(writers is { Length: > 0 });
+                foreach (TextWriter writer in writers)
+                {
+                    ArgumentNullException.ThrowIfNull(writer, nameof(writers));
+                }
+
+                _writers = writers;
+            }
+
+            public override Encoding Encoding => _writers[0].Encoding;
```
Comment (huoyaoyuan) [inline src/libraries/System.Private.CoreLib/src/System/IO/TextWriter.cs:1068]: Is this meaningful?

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/IO/TextWriter.cs:1068]: Can you clarify your question? What would you expect to happen instead? At least by specifying it uses the encoding from the first, someone can choose to order the writers appropriately. If the encodings differ between the writers and the consumer actually uses the encoding, there's no value here that would "just work".

Comment (huoyaoyuan) [reply src/libraries/System.Private.CoreLib/src/System/IO/TextWriter.cs:1068]: My question is how the get-only `Encoding` property is supposed to be used. It seems that no functionality of the broadcasting writer will be affected.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/IO/TextWriter.cs:1068]: It's a very rarely used property that a consumer can use if they need to generate bytes in the same encoding that the writer will use.

**Comment by stephentoub** [comment]
> For this behavior, Chain seems to be a better name for me.

When I think Chain, I think of it as first writing everything to the first writer, and then when the first writer is full, writing everything to the second writer, etc., rather than writing everything to everything.

> And for the name of broadcasting I would expect the behavior of broadcasting.WriteAsync to be something like Task.WhenAll.

Why? Does that mean you'd expect broadcasting.Write would use Parallel.For across all of the writers? I don't see why broadcast implies concurrent execution.

**Comment by skyoxZ** [comment]
> Why? Does that mean you'd expect broadcasting.Write would use Parallel.For across all of the writers? I don't see why broadcast implies concurrent execution.

My intuition tells me that a sync method usually doesn't contain any concurrent operation but it's normal if an async method contains some concurrent operations. Then considering the name of Broadcast, I would be a little surprised its `WriteAsync` doesn't implement concurrent so the operation takes more time to finish than I expected.

**Comment by stephentoub** [comment]
> but it's normal if an async method contains some concurrent operations

It's exceedingly rare for methods in the core libraries to introduce any concurrency as part of their operation (eg fork multiple tasks and then join with them). The only one I can think of off the top of my head is in Microsoft. Extensions and is opt-in.

File: src/libraries/System.Runtime/tests/System.IO.Tests/TextWriter/TextWriterTests.cs [hunk 2]
```diff
@@ -685,6 +685,186 @@ public async Task FlushAsync_Precanceled()
...
+            await oracle.WriteLineAsync((StringBuilder)null);
+            await broadcasting.WriteLineAsync((StringBuilder)null);
+            await oracle.WriteLineAsync(new StringBuilder("Wednesday"));
+            await broadcasting.WriteLineAsync(new StringBuilder("Wednesday"));
+
+            string expected = oracle.ToString();
+            Assert.Equal(expected, sw1.ToString());
+            Assert.Equal(expected, sw2.ToString());
+        }
+
```
Comment (adamsitnik) [inline src/libraries/System.Runtime/tests/System.IO.Tests/TextWriter/TextWriterTests.cs:867]: Could you please add a test that ensures that `BroadcastingTextWriter` overrides all virtual methods of `TextWriter`? I would like to make sure that we never miss it when adding new `virtual` methods to `TextWrtier`.

File: src/libraries/System.Runtime/tests/System.IO.Tests/TextWriter/TextWriterTests.cs [hunk 3]
```diff
@@ -685,6 +685,186 @@ public async Task FlushAsync_Precanceled()
             Assert.Equal(cts.Token, (await Assert.ThrowsAnyAsync<OperationCanceledException>(() => t)).CancellationToken);
         }
 
+        [Fact]
+        public async Task CreateBroadcasting_DelegatesToAllWriters()
+        {
+            AssertExtensions.Throws<ArgumentNullException>("writers", () => TextWriter.CreateBroadcasting(null));
+            AssertExtensions.Throws<ArgumentNullException>("writers", () => TextWriter.CreateBroadcasting(new TextWriter[] { null }));
+            AssertExtensions.Throws<ArgumentNullException>("writers", () => TextWriter.CreateBroadcasting(new TextWriter[] { new StringWriter(), null }));
+            AssertExtensions.Throws<ArgumentNullException>("writers", () => TextWriter.CreateBroadcasting(new TextWriter[] { null, new StringWriter() }));
```
Comment (adamsitnik) [inline src/libraries/System.Runtime/tests/System.IO.Tests/TextWriter/TextWriterTests.cs:694]: please move the input verification to a dedicated test

**Review by adamsitnik — DISMISSED** [review]
It looks good, but it needs some minor polishing before we merge it. Thank you @stephentoub !

PS. I am using "Request changes" on purpose, as agreed in our Stale PRs discussion. To make others start using it and to treat internal and external PRs equally.

File: src/libraries/System.Private.CoreLib/src/System/IO/TextWriter.cs [hunk 4]
```diff
@@ -1018,5 +1021,508 @@ public override Task FlushAsync(CancellationToken cancellationToken)
...
+        /// <para>
+        /// <see cref="Encoding"/> will return the corresponding object from first writer in <paramref name="writers"/>.
+        /// </para>
+        /// </remarks>
+        public static TextWriter CreateBroadcasting(params TextWriter[] writers)
+        {
+            ArgumentNullException.ThrowIfNull(writers);
+
+            return writers.Length != 0 ?
+                new BroadcastingTextWriter([.. writers]) :
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/IO/TextWriter.cs:1049]: The [ ] part is a collection expression that will create whatever is the target collection type; in this case, that's an array. The .. part is a "spread", which means it takes all the elements of the underlying collection and puts them into the collection as if you'd done so one by one.  How the compiler does so is an implementation detail, and it has various approaches to try to make it as efficient as possible (with the freedom to add more optimizations in the future). This particular expression will compile down into wrapping the array in a span and calling ToArray on the span.

**Comment by stephentoub** [comment]
@adamsitnik, I addressed your feedback. Please re-review.

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @stephentoub !

**Pull request merged**